core/src/main/java/com/datastax/oss/driver/internal/core/os/GraalLibc.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core.os;
+
+import java.util.Optional;
+
+public class GraalLibc implements Libc {
+
+  @Override
+  public boolean available() {
+    return true;
+  }
+
+  /* Substrate includes a substitution for Linux + Darwin which redefines System.nanoTime() to use
+   * gettimeofday() (unless platform-specific higher-res clocks are available, which is even better). */
+  @Override
+  public Optional<Long> gettimeofday() {
+    return Optional.of(Math.round(System.nanoTime() / 1_000d));
```

Comment:
`gettimeofday()` has microsecond resolution. If `nanoTime()` uses it, it likely returns a multiple of 1000. So an integer division should be sufficient.

Reply:
gettimeofday() does indeed have microsecond resolution, but note the comment above about trying to use the higher resolution Linux clocks if they are available.  That's the clock_gettime() call in https://github.com/oracle/graal/blob/master/substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/linux/LinuxSubstitutions.java#L47-L51.  If that succeeds then we'd need rounding to get to microseconds.

Reply:
I had missed that, makes sense then 👍

Reply:
1. What happens if Graal changes its substitution for `System.nanoTime()` to use something that is not a wall-clock time? It looks to me that we are relying on implementation details that could change later.
2. What happens for platforms that do not have a similar substitution, Windows for example? Is this method going to return the actual call to `System.nanoTime()`?

Reply:
Both legit points @adutra.  We've talked about this offline but I'll summarize those conversations here for the record.

Re: relying on Graal's substitution impl... I wrestled with this as well, and I don't love the idea of relying on a third party like this.  In practice, though, the logic in the substitution actually _improves on_ just relying on gettimeofday() and what we'd really be worried about here is if the Graal team made a change in the future which somehow _lessened_ the accuracy of their substitution.  This is certainly not impossible, but it seems fairly unlikely.

What we have right now seems "good enough" for an initial iteration.  We discussed filing a follow-up ticket to explore this in the future in order to see if it's worthwhile to remove this "dependency" [1] on the substitution.  I've filed JAVA-2768 to consider this change.

Re: platforms without the substitution... it seems to make sense to make the Native.isXAvailable() checks return a result based on platform in order to account for this.  If the platform is observed to be LInux or OS X we return true, otherwise false.  This approach has the advantage of playing nicely with how the rest of the driver codebase uses Native.

[1] "dependency" is (sort of) the wrong phrase here.  We're relying on System.nanoTime() doing the right thing, not any additional dependency.  Hypothetically any JDK version could also change this behaviour in it's internals... which in turn actually supports the idea of calling getimeofday() directly in order to avoid reliance on volatile parts of the platform.